File: backend/app/services/websocket.py
from fastapi import WebSocket
from typing import Dict, List
import json
from datetime import datetime

class WebSocketService:
    def __init__(self):
        self.active_connections: Dict[int, WebSocket] = {}
        self.typing_users: Dict[int, Dict[int, datetime]] = {}  # room_id -> {user_id -> timestamp}

    async def connect(self, websocket: WebSocket, user_id: int):
        await websocket.accept()
        self.active_connections[user_id] = websocket

    def disconnect(self, user_id: int):
        if user_id in self.active_connections:
            del self.active_connections[user_id]

    async def send_message(self, user_id: int, message: dict):
        if user_id in self.active_connections:
            await self.active_connections[user_id].send_json(message)

    async def broadcast_message(self, room_id: int, message: dict):
        for user_id, websocket in self.active_connections.items():
            if user_id in self.typing_users.get(room_id, {}):
                await websocket.send_json(message)

    def update_typing_status(self, room_id: int, user_id: int):
        if room_id not in self.typing_users:
            self.typing_users[room_id] = {}
        self.typing_users[room_id][user_id] = datetime.utcnow()

    def remove_typing_status(self, room_id: int, user_id: int):
        if room_id in self.typing_users and user_id in self.typing_users[room_id]:
            del self.typing_users[room_id][user_id]

    async def send_notification(self, user_id: int, notification: dict):
        """发送通知给指定用户"""
        message = {
            "type": "notification",
            "data": notification
        }
        await self.send_message(user_id, message)

    async def broadcast_notification(self, user_ids: List[int], notification: dict):
        """广播通知给多个用户"""
        message = {
            "type": "notification",
            "data": notification
        }
        for user_id in user_ids:
            if user_id in self.active_connections:
                await self.active_connections[user_id].send_json(message)

websocket_service = WebSocketService() 

File: frontend/src/services/websocket.ts
import { Message, WebSocketMessage } from '../types';

export class WebSocketService {
    private ws: WebSocket | null = null;
    private reconnectAttempts = 0;
    private maxReconnectAttempts = 5;
    private reconnectTimeout = 3000;
    private messageHandlers: ((message: Message) => void)[] = [];
    private notificationHandlers: ((notification: any) => void)[] = [];
    private typingHandlers: ((data: { room_id: number; user_id: number; is_typing: boolean }) => void)[] = [];

    constructor(private token: string) {}

    connect() {
        const wsUrl = `${process.env.REACT_APP_WS_URL}/ws?token=${this.token}`;
        this.ws = new WebSocket(wsUrl);

        this.ws.onopen = () => {
            console.log('WebSocket connected');
            this.reconnectAttempts = 0;
        };

        this.ws.onmessage = (event) => {
            const data = JSON.parse(event.data);
            
            switch (data.type) {
                case 'message':
                    this.messageHandlers.forEach(handler => handler(data.data));
                    break;
                case 'notification':
                    this.notificationHandlers.forEach(handler => handler(data.data));
                    break;
                case 'typing':
                    this.typingHandlers.forEach(handler => handler(data.data));
                    break;
            }
        };

        this.ws.onclose = () => {
            console.log('WebSocket disconnected');
            this.handleReconnect();
        };

        this.ws.onerror = (error) => {
            console.error('WebSocket error:', error);
        };
    }

    private handleReconnect() {
        if (this.reconnectAttempts < this.maxReconnectAttempts) {
            this.reconnectAttempts++;
            setTimeout(() => {
                console.log(`Attempting to reconnect (${this.reconnectAttempts}/${this.maxReconnectAttempts})`);
                this.connect();
            }, this.reconnectTimeout);
        }
    }

    disconnect() {
        if (this.ws) {
            this.ws.close();
            this.ws = null;
        }
    }

    sendMessage(content: string, type: 'text' | 'image') {
        if (this.ws?.readyState === WebSocket.OPEN) {
            this.ws.send(JSON.stringify({
                type: 'message',
                data: {
                    content,
                    message_type: type
                }
            }));
        }
    }

    sendTyping(isTyping: boolean) {
        if (this.ws?.readyState === WebSocket.OPEN) {
            this.ws.send(JSON.stringify({
                type: 'typing',
                data: {
                    is_typing: isTyping
                }
            }));
        }
    }

    addTypingHandler(handler: (data: { room_id: number; user_id: number; is_typing: boolean }) => void) {
        this.typingHandlers.push(handler);
        return () => {
            this.typingHandlers = this.typingHandlers.filter(h => h !== handler);
        };
    }

    onMessage(handler: (message: Message) => void) {
        this.messageHandlers.push(handler);
        return () => {
            this.messageHandlers = this.messageHandlers.filter(h => h !== handler);
        };
    }

    onNotification(handler: (notification: any) => void) {
        this.notificationHandlers.push(handler);
        return () => {
            this.notificationHandlers = this.notificationHandlers.filter(h => h !== handler);
        };
    }

    onTyping(handler: (data: { room_id: number; user_id: number; is_typing: boolean }) => void) {
        this.typingHandlers.push(handler);
        return () => {
            this.typingHandlers = this.typingHandlers.filter(h => h !== handler);
        };
    }
} 